x/slashing/keeper/migrations.go

```diff
@@ -118,7 +123,92 @@ func (m Migrator) Migrate2to3(ctx sdk.Context) error {
 		if err != nil {
 			return err
 		}
-		m.keeper.SetValidatorMissedBlocks(ctx, consAddrKey, missedBlockArray)
+		bz := m.keeper.cdc.MustMarshal(&missedBlockArray)
+		store.Set(types.ValidatorMissedBlockBitArrayKey(consAddrKey), bz)
+	}
+	ctx.Logger().Info("Done migrating")
+	return nil
+}
+
+// Migrate3to4 migrates from version 3 to 4.
+func (m Migrator) Migrate3to4(ctx sdk.Context) error {
+	ctx.Logger().Info("Migrating 3 -> 4")
+	store := ctx.KVStore(m.keeper.storeKey)
+	valMissedMap := make(map[string]types.ValidatorMissedBlockArray)
+	ctx.Logger().Info("Migrating Signing Info")
+	signInfoIter := sdk.KVStorePrefixIterator(store, types.ValidatorSigningInfoKeyPrefix)
+	newSignInfoKeys := [][]byte{}
+	newSignInfoVals := []types.ValidatorSigningInfo{}
+	// use previous height to calculate index offset
+	window := m.keeper.SignedBlocksWindow(ctx)
+	index := window - 1
+	defer signInfoIter.Close()
+	for ; signInfoIter.Valid(); signInfoIter.Next() {
+		ctx.Logger().Info(fmt.Sprintf("Migrating Signing Info for key: %v\n", signInfoIter.Key()))
+		var oldInfo types.ValidatorSigningInfoLegacyMissedHeights
+		m.keeper.cdc.MustUnmarshal(signInfoIter.Value(), &oldInfo)
+
+		newInfo := types.ValidatorSigningInfo{
+			Address:             oldInfo.Address,
+			StartHeight:         oldInfo.StartHeight,
+			IndexOffset:         index,
+			JailedUntil:         oldInfo.JailedUntil,
+			Tombstoned:          oldInfo.Tombstoned,
+			MissedBlocksCounter: oldInfo.MissedBlocksCounter,
+		}
+		newSignInfoKeys = append(newSignInfoKeys, signInfoIter.Key())
+		newSignInfoVals = append(newSignInfoVals, newInfo)
+	}
+	signInfoIter.Close()
```

Comment:
we're closing the iterator twice - is that intentional?

Reply:
yeah, @yzang2019 said this is fine. We're closing this one here to make sure the next one being opened is done safely, and the defer for close is in case theres an error while iterating